### PR TITLE
Convert TraceProducer into an interface

### PR DESCRIPTION
**Pull request by zenios**
How a trace producer works is an implementation detail and should not be enforced by using an abstract class.

This is how i use it in my own application and vertx

```
public class ScheduledBatchingTraceProducer implements TraceProducer {
    private final Logger logger = LoggerFactory.getLogger(ScheduledBatchingTraceProducer.class);
    private final Consumer<Reports.ReportHeader.Builder> customizeReportHeader;
    private final Uploader uploader;
    private final BlockingQueue<Reports.Trace> queue;
    private final Vertx vertx;
    private final long timerId;
    private boolean stopped = false;

    public static Builder newBuilder() {
        return new Builder();
    }

    public ScheduledBatchingTraceProducer(Consumer<Reports.ReportHeader.Builder> customizeReportHeader,
                                          Uploader uploader,
                                          Vertx vertx,
                                          Duration batchingWindow,
                                          BlockingQueue<Reports.Trace> queue) {
        this.customizeReportHeader = customizeReportHeader;
        this.uploader = uploader;
        this.queue = queue;
        this.vertx = vertx;
        this.timerId = vertx.setPeriodic(batchingWindow.toMillis(), event -> safePerform());
    }

    @Override
    public void submit(Reports.Trace trace) {
        if (stopped) {
            logger.warn("Producer is stopped.Submit not accepted");
            return;
        }

        try {
            queue.add(trace);
        } catch (Exception e) {
            logger.error("Error submitting to queue", e);
        }
    }

    public Future<Void> shutdown() {
        stopped = true;
        vertx.cancelTimer(timerId);
        return safePerform();
    }

    private Future<Void> safePerform() {
        try {
            return perform();
        } catch (Exception e) {
            logger.error("Error during perform", e);
            return Future.failedFuture(e);
        }
    }

    private Future<Void> perform() {
        List<Reports.Trace> traces = new ArrayList<>();
        queue.drainTo(traces);
        if (traces.isEmpty())
            return Future.succeededFuture();

        return uploader.upload(reportBuilder.build(traces, customizeReportHeader));
    }

    public static class Builder {
        private Consumer<Reports.ReportHeader.Builder> _customizeReportHeader;
        private Uploader _uploader;
        private Vertx _vertx;
        private Duration _batchingWindow = Duration.ofSeconds(10);
        private BlockingQueue<Reports.Trace> _queue = new ArrayBlockingQueue<>(4096);

        public ScheduledBatchingTraceProducer build() {
            return new ScheduledBatchingTraceProducer(_customizeReportHeader, _uploader, _vertx, _batchingWindow, _queue);
        }

        public Builder customizeHeader(Consumer<Reports.ReportHeader.Builder> customizeReportHeader) {
            this._customizeReportHeader = customizeReportHeader;
            return this;
        }

        public Builder uploader(Uploader uploader) {
            this._uploader = uploader;
            return this;
        }

        public Builder batchingWindow(Duration batchingWindow) {
            this._batchingWindow = batchingWindow;
            return this;
        }

        public Builder queue(BlockingQueue<Reports.Trace> queue) {
            this._queue = queue;
            return this;
        }

        public Builder vertx(Vertx vertx) {
            this._vertx = vertx;
            return this;
        }
    }
}
```

You can see that shutdown does not block waiting for an executor to complete. It returns a future and let the caller decide what to do

### CODE REVIEW

File: src/main/java/com/braintreepayments/apollo_tracing_uploader/TraceProducer.java [hunk 1]
```diff
@@ -1,18 +1,10 @@
...
-public abstract class TraceProducer {
-  protected final FullTracesReportBuilder reportBuilder = new FullTracesReportBuilder();
-  protected final Consumer<Reports.ReportHeader.Builder> customizeReportHeader;
-  protected final Uploader uploader;
-
-  public TraceProducer(Consumer<Reports.ReportHeader.Builder> customizeReportHeader, Uploader uploader) {
-    this.customizeReportHeader = customizeReportHeader;
-    this.uploader = uploader;
-  }
+public interface TraceProducer {
```
Comment (alexgenco) [inline src/main/java/com/braintreepayments/apollo_tracing_uploader/TraceProducer.java:5]: I do agree it feels a little awkward for this to be an abstract class, but the reason I did it was to enforce the need for a report header customizer consumer and an uploader in the constructor, because any "implementation" of `TraceProducer` is responsible for calling the consumer on the report header and uploading the report.

Would it make sense to add the interface as you've done, but then provide an `AbstractTraceProducer` implementation that contains this constructor? Then `ScheduledBatchingTraceProducer` could `extend AbstractTraceProducer`?

File: src/main/java/com/braintreepayments/apollo_tracing_uploader/impl/Uploader.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.braintreepayments.apollo_tracing_uploader.impl;
```
Comment (alexgenco) [inline src/main/java/com/braintreepayments/apollo_tracing_uploader/impl/Uploader.java:1]: Why did you move this to the `impl` package? This is meant for concrete implementations of interfaces.

**Comment by zenios** [comment]
> Originally I think you suggested `Uploader::upload` return a future, is that no longer needed?

If you can see at my implementation, safePerform is returning a future. Instead of modifying upload interface to return a CompletableFuture i decided its better to consider uploading an implementation detail and let the implementor handle it they way it wants.

> I do agree it feels a little awkward for this to be an abstract class, but the reason I did it was to enforce the need for a report header customizer consumer and an uploader in the constructor, because any "implementation" of TraceProducer is responsible for calling the consumer on the report header and uploading the report.

>Why did you move this to the impl package? This is meant for concrete implementations of interfaces. 

As said what will happen after a report is submitted should be considered an implementation detail and any assumptions made should be avoided.

I might want to create one class that calls the consumer and uploads (All in the same class). In that case what will Uploader provide. I know its a bad coding practice but its an example to show why Uploader interface is moved into impl

> Would it make sense to add the interface as you've done, but then provide an AbstractTraceProducer implementation that contains this constructor? Then ScheduledBatchingTraceProducer could extend AbstractTraceProducer?

I am not against that. It will help in cases you want a producer that it almost the same with the one provided

> Also, is your code snippet on the PR description going to be a part of this change as well, or just an example of how one could implement the proposed TraceProducer interface?

The code snippet provided is my implementation of TraceProducer using vertx. The one i use in production right now

Concluding i would like to say that this is only my opinion. **Thanks for your time**

**Comment by alexgenco** [comment]
I think I would prefer leaving the uploader where it is and adding the `AbstractTraceProducer` abstract class (probably in the `impl` package). I think of the `impl` package as simply "classes that implement the interfaces provided in this library", not necessarily things that are implementation details.

How does that sound? This still supports your vertx implementation right?

**Comment by zenios** [comment]
Sure.no problem

**Comment by zenios** [comment]
As long as trace producer is an interface then i am fine with that

**Comment by alexgenco** [comment]
just opened a PR for my suggestions https://github.com/zenios/apollo-tracing-uploader-java/pull/1

**Comment by zenios** [comment]
> Great, feel free to merge this and I'll merge the main PR.

Done

**Comment by alexgenco** [comment]
Thanks @zenios !

**Comment by zenios** [comment]
Any chance of a release publish?

**Comment by alexgenco** [comment]
@zenios just published 0.2.0
